### PR TITLE
Noisy layer

### PR DESCRIPTION
**Pull request by jvmncs**
Implementation of Noisy Networks per #2024.  The gist link from that issue is now obsolete; the forward pass no longer resamples the noise tensors each time, and I've added a method `reset_noise` to resample the noise tensors.  Also, I used `self.training` to differentiate between train and eval passes.  I ran some basic tests to make sure methods were functioning, but I still need to do more testing.  Also, I'm not sure how to edit the docs. If someone can point me in the right direction for expectations on writing docs, I'd appreciate it.

### CODE REVIEW

File: torch/nn/modules/linear.py [hunk 1]
```diff
@@ -12,16 +13,14 @@ class Linear(Module):
     Args:
         in_features: size of each input sample
         out_features: size of each output sample
-        bias: If set to False, the layer will not learn an additive bias.
-            Default: True
+        bias: If set to False, the layer will not learn an additive bias. Default: True
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: I think we're trying to keep the line lengths under 80 chars, so please undo these formatting changes (I also agree they look nicer but hey).

File: torch/nn/modules/linear.py [hunk 2]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+    More details can be found in the paper `Noisy Networks for Exploration` _ .
+    Args:
+        in_features: size of each input sample
+        out_features: size of each output sample
+        bias: If set to False, the layer will not learn an additive bias.
+            Default: True
+        factorised: whether or not to use factorised noise.
+            Default: True
+        std_init: initialization constant for standard deviation component of
+            weights. If None, defaults to 0.017 for independent and 0.4 for
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: Might want to add `(weight_sigma and bias_sigma)` or something similar to make it clearer (that it isn't something like the std dev of noise used to init all weights or something).

File: torch/nn/modules/linear.py [hunk 3]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+        else:
+            self.register_parameter('bias', None)
+        if not std_init:
+            if self.factorised:
+                self.std_init = 0.4
+            else:
+                self.std_init = 0.017
+        else:
+            self.std_init = std_init
+        self.reset_parameters(bias)
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: Could you add `self.bias = bias` above and therefore run `self.reset_parameters()` without any arguments (same as `nn.Linear`)?

File: torch/nn/modules/linear.py [hunk 4]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+        else:
+            mu_range = math.sqrt(3. / self.weight_mu.size(1))
+            self.weight_mu.data.uniform_(-mu_range, mu_range)
+            self.weight_sigma.data.fill_(self.std_init)
+            if bias:
+                self.bias_mu.data.uniform_(-mu_range, mu_range)
+                self.bias_sigma.data.fill_(self.std_init)
+
+    def scale_noise(self, size):
+        x = torch.Tensor(size).normal_()
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: You can just use `x = torch.randn(size)` to get a N(0, 1) tensor straight away.

File: torch/nn/modules/linear.py [hunk 5]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+        return x
+
+    def reset_noise(self):
+        if self.factorised:
+            epsilon_in = self.scale_noise(self.in_features)
+            epsilon_out = self.scale_noise(self.out_features)
+            self.weight_epsilon = Variable(epsilon_out.ger(epsilon_in))
+            self.bias_epsilon = Variable(self.scale_noise(self.out_features))
+        else:
+            self.weight_epsilon = Variable(torch.Tensor((self.out_features, self.in_features)).normal_())
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: See comment above for both of these too.

File: torch/nn/modules/linear.py [hunk 6]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
             + ', in2_features=' + str(self.in2_features) \
             + ', out_features=' + str(self.out_features) + ')'
 
+
+class NoisyLinear(Module):
+    """Applies a noisy linear transformation to the incoming data:
+        :math:`y = (mu_w + sigma_w \cdot epsilon_w)x
+            + mu_b + sigma_b \cdot epsilon_b`
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: Add a note in the docs to note the equation used during evaluation (just the mu's).

File: torch/nn/modules/linear.py [hunk 7]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+            return F.linear(input,
+                            self.weight_mu + self.weight_sigma.mul(self.weight_epsilon),
+                            self.bias_mu + self.bias_sigma.mul(self.bias_epsilon))
+        else:
+            return F.linear(input, self.weight_mu, self.bias_mu)
+
+    def __repr__(self):
+        return self.__class__.__name__ + ' (' \
+            + str(self.in_features) + ' -> ' \
+            + str(self.out_features) + ')'
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: It's probably worth including `self.factorised` in this as well. `self.bias` is left out of `nn.Linear`, and `self.std_init` is only applicable at init.

File: torch/nn/modules/linear.py [hunk 8]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+            weights. If None, defaults to 0.017 for independent and 0.4 for
+            factorised. Default: None
+    Shape:
+        - Input: :math:`(N, in\_features)`
+        - Output: :math:`(N, out\_features)`
+    Attributes:
+        weight: the learnable weights of the module of shape
+            (out_features x in_features)
+        bias:   the learnable bias of the module of shape (out_features)
+    Examples::
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: The docs also need a reference to `reset_noise()` so that people know that they can call this. If you can't find any other modules with methods, just ping Soumith to find out what's appropriate.

File: torch/nn/modules/linear.py [hunk 9]
```diff
@@ -115,4 +116,99 @@ def __repr__(self):
...
+                self.bias_sigma.data.fill_(self.std_init / math.sqrt(self.bias_sigma.size(0)))
+        else:
+            mu_range = math.sqrt(3. / self.weight_mu.size(1))
+            self.weight_mu.data.uniform_(-mu_range, mu_range)
+            self.weight_sigma.data.fill_(self.std_init)
+            if bias:
+                self.bias_mu.data.uniform_(-mu_range, mu_range)
+                self.bias_sigma.data.fill_(self.std_init)
+
+    def scale_noise(self, size):
```
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: Might be good to rename this to `_scale_noise` as there is no need for users to call this.

**Comment by jvmancuso** [comment]
@soumith is the method documentation at lines 145/6 in a satisfactory format?  Can't find a module with a similar example.

**Comment by Kaixhin** [comment]
Everything looks good to me now :+1: Not sure about how best to document methods - pinging @alykhantejani for advice and what to do for testing.

**Comment by alykhantejani** [comment]
@jvmancuso I think you just need to add some test cases to the list here: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3114

you can follow the style of Linear here: https://github.com/pytorch/pytorch/blob/master/test/common_nn.py#L28

**Comment by alykhantejani** [comment]
@jvmancuso as for the docstring, you should prefix the `"""` with an `r` i.e. 
```python
r"""
   this is my docstring which is now treated as a raw string.
   This means escape characters like \ wont be converted and instead
   treated as a literal \
"""
```

File: torch/nn/modules/linear.py [hunk 10]
```diff
@@ -115,4 +116,109 @@ def __repr__(self):
             + ', in2_features=' + str(self.in2_features) \
             + ', out_features=' + str(self.out_features) + ')'
 
+
+class NoisyLinear(Module):
+    """Applies a noisy linear transformation to the incoming data.
+    During training:
+        :math:`y = (mu_w + sigma_w \cdot epsilon_w)x
```
Comment (alykhantejani) [inline torch/nn/modules/linear.py]: I believe this is inline math, if that's what you want. If you wanted it on a new line I **think** the syntax is 
```python
.. math:: <your math here>
```

But perhaps Soumith can confirm

File: torch/nn/modules/linear.py [hunk 11]
```diff
@@ -115,4 +116,109 @@ def __repr__(self):
...
+    Examples::
+        >>> m = nn.NoisyLinear(20, 30)
+        >>> input = autograd.Variable(torch.randn(128, 20))
+        >>> output = m(input)
+        >>> m.reset_noise()
+        >>> output_new = m(input)
+        >>> print(output)
+        >>> print(output_new)
+    """
+    def __init__(self, in_features, out_features, bias=True, factorised=True, std_init=None):
```
Comment (alykhantejani) [inline torch/nn/modules/linear.py]: Should we be using the US English `factorized` (with a `z`)?

File: torch/nn/modules/linear.py (hunk 10 above)
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: [Sphinx maths docs](http://www.sphinx-doc.org/en/stable/ext/math.html). So yes Aly is right.

File: torch/nn/modules/linear.py (hunk 11 above)
Comment (Kaixhin) [inline torch/nn/modules/linear.py]: For consistency yes...

**Comment by jvmancuso** [comment]
@alykhantejani pretty sure I fixed the math line in the docs at L123.  

I'm not sure that a test case like Linear's will work in this instance, since the output of the layer will be different for each forward pass by definition.  Specifically, there is no stable `reference_fn` to use, and I'm not sure how to test the functionality without a custom TestNN object that checks for AssertNotEqual instead of AssertEqual.

**Comment by alykhantejani** [comment]
@jvmancuso You don't have to add the `reference_fn` field but adding the entry to the dict will make sure the jacobian checks are done (numerical vs analytical)

**Comment by Kaixhin** [comment]
@alykhantejani Apart from the dict tests I think 2 custom tests would be good:

- In training mode, show that output 1, given a fixed input, is reasonably different from output 2, given the same input, but after `reset_noise` has been called. There is already an `assertNotEqual` function available to use.
- In evaluation mode, the output matches a linear layer with the same weights and biases.

It is pretty critical that the module shows these behaviours beyond simply passing the automatic differentiation checks.

**Comment by alykhantejani** [comment]
@Kaixhin agreed, these can just be added as regular test functions in `test_nn.py`

**Comment by jvmancuso** [comment]
Dug into these most recent failures.  Question: what do I do about test_noncontig?  It appears to be checking that using deepcopy to copy the layer doesn't change the gradient of the parameters.  I think deepcopy might be resampling the noise somewhere, which would definitely trigger assertEqual to fail on parameter grads if that's the case.  Can somebody else please take a look at this?

Also, no idea why test_Conv2d_backward_twice is failing.  Nothing I've done changes the Conv2d module or that test case, and I don't see how any changes I've made would cause it to fail.

**Comment by alykhantejani** [comment]
@jvmancuso sorry I've been away on vacation. Will try to take a look at this, this week.

File: test/test_nn.py [hunk 12]
```diff
@@ -2344,62 +2274,6 @@ def test_bce_with_logits_gives_same_result_as_sigmoid_and_bce_loss(self):
         weight = torch.rand(4)
         self.assertEqual(nn.BCEWithLogitsLoss(weight)(output, target), nn.BCELoss(weight)(sigmoid(output), target))
 
-        target = Variable(torch.FloatTensor(4, 1).fill_(0))
```
Comment (alykhantejani) [inline test/test_nn.py]: @jvmancuso do you know why all of these are showing as deleted?

**Comment by alykhantejani** [comment]
@jvmancuso The issue with `test_noncontig` is because the function tries to zero the gradients of the params [here](https://github.com/pytorch/pytorch/blob/f484a5fee8ee30d51903f99b5f4c0ce212ca0cfa/test/common_nn.py#L665) which calls this [snippet of code](https://github.com/pytorch/pytorch/blob/977f9644c08680df69df108b037cdd5ee9f47f94/test/test_nn.py#L256) which only zeros gradients for `weight` and `bias` and not the other params in this module (`weight_mu` etc.)

I'm not quite sure why `_zero_grad_parameters` explicitly names `weight` and `bias`, but perhaps this function can be replaced with a call to `module. zero_grad ()`. Although there is a `detach()` call in `_zero_grad_parameters`, so if this is actually needed this function can instead loop through the modules parameters and manually zero the grads and detach.

@soumith @apaszke wdyt?

@jvmancuso in terms of the other test failing, try and pull in changes from upstream/master as you currently have merge conflicts anyway.

**Comment by jvmancuso** [comment]
Hi all, it's been awhile, and I haven't forgotten about this.  I'm doing an analysis of DeepMind's Noisy Nets versus OpenAI's Parameter Space Noise, and wanted to revisit this when that's done.  I'll be implementing the OpenAI version shortly, and wanted to investigate including that in this PR.  I'll close this for now and reopen when I've thought through that a bit more.

**Comment by Kaixhin** [comment]
@jvmancuso for future reference, this doesn't work with CUDA. I think the best solution is to register `weight_epsilon` and `bias_epsilon` as buffers so that when the model is cast to CUDA they are cast as well - and then the generated noise needs to be copied over. You can see an example [here](https://github.com/Kaixhin/Rainbow/blob/master/model.py#L9-L46).

**Comment by jvmancuso** [comment]
@Kaixhin I had made a few changes to the code to accommodate for that but hadn't committed them.  Your solution is more elegant though, I'll integrate it into my work.  Thanks!
